## Assignments
In this assignment, you'll continue working with the Credit Card Fraud Detection dataset from Kaggle. Load the dataset before working on the following tasks.

https://www.kaggle.com/mlg-ulb/creditcardfraud/download


To complete this assignment, submit your solutions to the following tasks as a link to your Jupyter Notebook on GitHub.


1. How many transactions are there in total?

2. How many transactions are fraud, and how many are not fraud?

3. What is the maximum amount in fraud transactions?

Submit your work below. You can also take a look at this example solution.


https://drive.google.com/file/d/19bMnC6yJahYpmcECkZXFiFnb9UHbMwjT/view?usp=sharing

In [1]:
import warnings
warnings.filterwarnings("ignore")

from dask.distributed import Client, progress

client = Client(n_workers=4, threads_per_worker=2, memory_limit='2GB')
client

<Client: 'tcp://127.0.0.1:50985' processes=4 threads=8, memory=8.00 GB>

In [2]:
# DataFrames implement the pandas API
import dask.dataframe as dd

# This loads the data into a Dask DataFrame
df = dd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/creditcard.csv')

In [3]:
print(df)

Dask DataFrame Structure:
                Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount  Class
npartitions=3                                                                                                                                                                                                                                                                                   
               int64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  int64
                 ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      

In [4]:
df.dtypes

Time        int64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

In [5]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [6]:
df.describe()

Dask DataFrame Structure:
                  Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount    Class
npartitions=1                                                                                                                                                                                                                                                                                       
               float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...
Dask Name: describe-numeric, 344 tasks

In [7]:
df = dd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/creditcard.csv', dtype={'Time': 'float64'})

In [8]:
df.describe().compute()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.187934e-15  3.384974e-16 -1.328443e-15  2.075692e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     49346.000000 -7.868600e-01 -5.536365e-01 -7.156553e-01 -7.067841e-01   
50%     76029.000000  9.301532e-02  8.999754e-02  1.743346e-01  1.809941e-01   
75%    138472.000000  1.912886e+00  8.916958e-01  1.373682e+00  1.012508e+00   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.707851e-16  1.494498e-15 -5.588401e-16  1.149614e-16 -2.403811e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -4.093779e-01 -6.546632e-01 -4.787135e-01 -1.339191e-01 -5.477089e-01   
50%    1.480477e-01 -1.649056e-01  1.590763e-01  7.790153e-02  2.583745e-02   
75%    8.613822e-01  4.775488e-01  7.334868e-01  3.706835e-01  7.092123e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.644587e-16 -3.544643e-16  2.626548e-16  4.473016e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.256367e-01 -5.250530e-01 -1.301482e-01 -3.241623e-01   
50%    ...  6.636310e-03  1.241855e-01  5.160336e-02  6.794774e-02   
75%    ...  2.385322e-01  7.308589e-01  2.346538e-01  5.274699e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.141329e-16  1.683905e-15 -3.660403e-16 -1.221465e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -2.218416e-01 -2.821043e-01 -6.089022e-02 -2.762231e-02       6.840000   
50%    1.674311e-01 -1.290092e-03  1.088230e-02  2.345040e-02      24.990000   
75%    4.190202e-01  2.938470e-01  1.060637e-01  8.061929e-02      84.910000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

1. How many transactions are there in total?

In [15]:
len(df)

284807

In [16]:
a = df.shape
a[0].compute(),a[1]

(284807, 31)

There are 284,807 transactions (rows) in total.

2. How many transactions are fraud, and how many are not fraud?

In [18]:
fraud = df[df["Class"] ==1]
fraud

Dask DataFrame Structure:
                  Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount  Class
npartitions=3                                                                                                                                                                                                                                                                                     
               float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  int64
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
Dask Name: getitem, 12 tasks

In [19]:
fraud.compute()

Time        V1        V2        V3        V4        V5        V6  \
541       406.0 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623       472.0 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
4920     4462.0 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788   
6108     6986.0 -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536   
6329     7519.0  1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746   
...         ...       ...       ...       ...       ...       ...       ...   
38075  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
38355  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
38361  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
39356  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
39886  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

             V7        V8        V9  ...       V21       V22       V23  \
541   -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211   
623    0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
4920   0.562320 -0.399147 -0.238253  ... -0.294166 -0.932391  0.172726   
6108  -3.496197 -0.248778 -0.247768  ...  0.573574  0.176968 -0.436207   
6329   1.713445 -0.496358 -1.282858  ... -0.379068 -0.704181 -0.656805   
...         ...       ...       ...  ...       ...       ...       ...   
38075 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
38355 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
38361 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
39356 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
39886  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

            V24       V25       V26       V27       V28  Amount  Class  
541    0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
623   -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
4920  -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
6108  -0.053502  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
6329  -1.632653  1.488901  0.566797 -0.010016  0.146793    1.00      1  
...         ...       ...       ...       ...       ...     ...    ...  
38075 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
38355 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
38361  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
39356 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
39886 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[492 rows x 31 columns]

There are 492 fraud transactions.

In [20]:
notfraud = df[df["Class"] ==0]
notfraud

Dask DataFrame Structure:
                  Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount  Class
npartitions=3                                                                                                                                                                                                                                                                                     
               float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  int64
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
Dask Name: getitem, 12 tasks

In [21]:
notfraud.compute()

Time         V1         V2        V3        V4        V5        V6  \
0           0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321  0.462388   
1           0.0   1.191857   0.266151  0.166480  0.448154  0.060018 -0.082361   
2           1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198  1.800499   
3           1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4           2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193  0.095921   
...         ...        ...        ...       ...       ...       ...       ...   
43014  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473 -2.606837   
43015  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229  1.058415   
43016  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515  3.031260   
43017  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961  0.623708   
43018  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546 -0.649617   

             V7        V8        V9  ...       V21       V22       V23  \
0      0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1     -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2      0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3      0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4      0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
...         ...       ...       ...  ...       ...       ...       ...   
43014 -4.918215  7.305334  1.914428  ...  0.213454  0.111864  1.014480   
43015  0.024330  0.294869  0.584800  ...  0.214205  0.924384  0.012463   
43016 -0.296827  0.708417  0.432454  ...  0.232045  0.578229 -0.037501   
43017 -0.686180  0.679145  0.392087  ...  0.265245  0.800049 -0.163298   
43018  1.577006 -0.414650  0.486180  ...  0.261057  0.643078  0.376777   

            V24       V25       V26       V27       V28  Amount  Class  
0      0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1     -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      0  
2     -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3     -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4      0.141267 -0.206010  0.502292  0.219422  0.215153   69.99      0  
...         ...       ...       ...       ...       ...     ...    ...  
43014 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77      0  
43015 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79      0  
43016  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88      0  
43017  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00      0  
43018  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00      0  

[284315 rows x 31 columns]

There are 284315 not fraud transactions.

In [22]:
df["Class"].unique().compute()

0    0
1    1
Name: Class, dtype: int64

In [27]:
df.groupby("Class")["Time"].count().compute()

Class
0    284315
1       492
Name: Time, dtype: int64

3. What is the maximum amount in fraud transactions?

In [25]:
fraud.describe().compute()

Time          V1          V2          V3          V4  \
count     492.000000  492.000000  492.000000  492.000000  492.000000   
mean    80746.806911   -4.771948    3.623778   -7.033281    4.542029   
std     47835.365138    6.783687    4.291216    7.110937    2.873318   
min       406.000000  -30.552380   -8.402154  -31.103685   -1.313275   
25%     26556.000000   -3.434791    1.291527   -6.502294    2.572046   
50%     56806.000000   -0.902340    2.844795   -4.769752    4.284346   
75%    102622.000000    0.088907    5.262312   -1.601052    6.419365   
max    170348.000000    2.132386   22.057729    2.250210   12.114672   

               V5          V6          V7          V8          V9  ...  \
count  492.000000  492.000000  492.000000  492.000000  492.000000  ...   
mean    -3.151225   -1.397737   -5.568731    0.570636   -2.581123  ...   
std      5.372468    1.858124    7.206773    6.797831    2.500896  ...   
min    -22.105532   -6.406267  -43.557242  -41.044261  -13.434066  ...   
25%     -4.011781   -2.107219   -6.465488   -0.010874   -3.049294  ...   
50%     -0.940553   -1.256353   -2.187931    0.751883   -1.654109  ...   
75%      1.041801   -0.356259   -0.399869    1.942252   -0.779176  ...   
max     11.095089    6.474115    5.802537   20.007208    3.353525  ...   

              V21         V22         V23         V24         V25         V26  \
count  492.000000  492.000000  492.000000  492.000000  492.000000  492.000000   
mean     0.713588    0.014049   -0.040308   -0.105130    0.041449    0.051648   
std      3.869304    1.494602    1.579642    0.515577    0.797205    0.471679   
min    -22.797604   -8.887017  -19.254328   -2.028024   -4.781606   -1.152671   
25%      0.193607   -0.430388   -0.210422   -0.380075   -0.201183   -0.170881   
50%      0.642318    0.228484    0.013835   -0.023524    0.201196   -0.007772   
75%      1.719631    1.027445    0.357161    0.376769    0.634203    0.533343   
max     27.202839    8.361985    5.466230    1.091435    2.208209    2.745261   

              V27         V28       Amount  Class  
count  492.000000  492.000000   492.000000  492.0  
mean     0.170575    0.075667   122.211321    1.0  
std      1.376766    0.547291   256.683288    0.0  
min     -7.263482   -1.869290     0.000000    1.0  
25%      0.045917    0.000101     1.000000    1.0  
50%      0.551002    0.205958    11.390000    1.0  
75%      1.268958    0.409043   121.610000    1.0  
max      3.052358    1.779364  2125.870000    1.0  

[8 rows x 31 columns]

In [28]:
df[df.Class==1]["Amount"].max().compute()

2125.87

The maximum fraud transaction amount is 2125.87.

In [29]:
client.close()